placement/algo/algo.go

```diff
@@ -0,0 +1,482 @@
+					heap.Push(ph.hostHeap, triedHost)
+				}
+				break
+			}
+		}
+		if ph.hostHeap.Len() == 0 {
+			// this should only happen when RF > number of racks
+			return errNotEnoughRacks
+		}
+	}
+	return nil
+}
+
+// placeToRacksOtherThanOrigin move shards from a host to the rest of the cluster
+// the goal of this function is to assign "some" of the shards to the hosts in other racks
+func (ph placementHelper) placeToRacksOtherThanOrigin(shards map[uint32]struct{}, from *hostShards) {
+	var triedHosts []*hostShards
+	var sameRack []*hostShards
+outer:
+	for shard := range shards {
+		triedHosts = triedHosts[:0]
+		for ph.hostHeap.Len() > 0 {
+			tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+			// do not place to same rack for now
+			if from != nil && tryHost.hostRack() == from.hostRack() {
+				sameRack = append(sameRack, tryHost)
+				continue
+			}
+			triedHosts = append(triedHosts, tryHost)
+			if ph.hostHeap.getTargetLoadForHost(tryHost.hostAddress())-len(tryHost.shardsSet) <= 0 {
+				// this is where "some" is, at this point the best host option in the cluster
+				// from a different rack has reached its target load, time to break out of the loop
+				break outer
+			}
+			if ph.canAssignHost(shard, from, tryHost) {
+				ph.assignShardToHost(shard, tryHost)
+				delete(shards, shard)
+				for _, triedHost := range triedHosts {
+					heap.Push(ph.hostHeap, triedHost)
+				}
+				break
+			}
+		}
+	}
+	for _, host := range sameRack {
+		heap.Push(ph.hostHeap, host)
+	}
+	for _, triedHost := range triedHosts {
+		heap.Push(ph.hostHeap, triedHost)
+	}
+}
+
+func (ph placementHelper) generatePlacement() placementSnapshot {
+	return placementSnapshot{uniqueShards: ph.uniqueShards, rf: ph.rf, hostShards: ph.hostShards}
+}
+
+// hostHeap provides an easy way to get best candidate host to assign/steal a shard
+type hostHeap struct {
+	hosts                 []*hostShards
+	rackToHostsMap        map[string]map[*hostShards]struct{}
+	targetLoad            map[string]int
+	hostCapacityAscending bool
+}
+
+func newHostHeap(hosts []*hostShards, hostCapacityAscending bool, targetLoad map[string]int, rackToHostMap map[string]map[*hostShards]struct{}) *hostHeap {
+	hHeap := &hostHeap{hostCapacityAscending: hostCapacityAscending, hosts: hosts, targetLoad: targetLoad, rackToHostsMap: rackToHostMap}
+	heap.Init(hHeap)
+	return hHeap
+}
+
+func (hh hostHeap) getTargetLoadForHost(hostAddress string) int {
+	return hh.targetLoad[hostAddress]
+}
+func (hh hostHeap) Len() int {
+	return len(hh.hosts)
+}
+
+func (hh hostHeap) Less(i, j int) bool {
+	hostI := hh.hosts[i]
+	hostJ := hh.hosts[j]
+	leftLoadOnI := hh.getTargetLoadForHost(hostI.hostAddress()) - len(hostI.shardsSet)
+	leftLoadOnJ := hh.getTargetLoadForHost(hostJ.hostAddress()) - len(hostJ.shardsSet)
+	// if both host has tokens to be filled, prefer the one on a bigger rack
+	// since it tends to be more picky in accepting shards
+	if leftLoadOnI > 0 && leftLoadOnJ > 0 {
```

Comment:
yes, first of all, host will not go out of their target load, I only compare rack when both hosts have room left.
I was only comparing on token left on each host, but would end up with unbalanced distribution sometimes, for example:
replica 1: h1r1: [1], h2r1: [2], h3r2: [3], h4r3:[4]
now we add replica 2, 
if the adding order is 3,4,1,2. because we are looking at the potential capacity on each host, so all 4 hosts are at target - load = 2 - 1 = 1
so shard 3 might end up in h4, shard 4 might end up in h3, then when it comes to 1 and 2, they can only be added to h3 and h4 because of the rack check, thus an unbalanced distribution. This is what I meant about bigger rack being picky in accepting shards